templates/ros2_control/controller/test_dummy_chainable_controller.cpp

```diff
@@ -366,6 +372,166 @@ TEST_F(DummyClassNameTest, receive_message_and_publish_updated_status)
+  msg->joint_names = joint_names_;
+  msg->displacements.resize(joint_names_.size(), TEST_DISPLACEMENT);
+  msg->velocities.resize(joint_names_.size(), std::numeric_limits<double>::quiet_NaN());
+  msg->duration = std::numeric_limits<double>::quiet_NaN();
+  controller_->input_ref_.writeFromNonRT(msg);
+  const auto age_of_last_command =
+    controller_->get_node()->now() - (*(controller_->input_ref_.readFromNonRT()))->header.stamp;
+
+  ASSERT_FALSE(age_of_last_command <= controller_->ref_timeout_);
+  ASSERT_EQ((*(controller_->input_ref_.readFromRT()))->displacements[0], TEST_DISPLACEMENT);
+  ASSERT_EQ(
+    controller_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+  EXPECT_FALSE(std::isnan(joint_command_values_[0]));
+  ASSERT_NE((*(controller_->input_ref_.readFromRT()))->displacements[0], TEST_DISPLACEMENT);
+  EXPECT_EQ(joint_command_values_[STATE_MY_ITFS], TEST_DISPLACEMENT);
+  ASSERT_TRUE(std::isnan((*(controller_->input_ref_.readFromRT()))->displacements[0]));
+}
+
+TEST_F(
+  DummyClassNameTest,
+  when_ref_timeout_zero_for_reference_callback_expect_reference_msg_being_used_only_once)
+{
+  SetUpController();
+  rclcpp::executors::MultiThreadedExecutor executor;
+  executor.add_node(controller_->get_node()->get_node_base_interface());
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_EQ(controller_->on_activate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+
+  EXPECT_EQ((*(controller_->input_ref_.readFromNonRT()))->joint_names.size(), joint_names_.size());
+  EXPECT_EQ((*(controller_->input_ref_.readFromNonRT()))->joint_names[0], joint_names_[0]);
+  EXPECT_TRUE(std::isnan((*(controller_->input_ref_.readFromNonRT()))->displacements[0]));
+  EXPECT_TRUE(std::isnan((*(controller_->input_ref_.readFromNonRT()))->velocities[0]));
+  EXPECT_TRUE(std::isnan((*(controller_->input_ref_.readFromNonRT()))->duration));
+  controller_->ref_timeout_ = rclcpp::Duration::from_seconds(0.0);
+  // reference_callback() is called implicitly when publish_commands() is called.
+  publish_commands(controller_->get_node()->now());
+  ASSERT_TRUE(controller_->wait_for_commands(executor));
+  EXPECT_EQ((*(controller_->input_ref_.readFromNonRT()))->joint_names.size(), joint_names_.size());
+  EXPECT_EQ((*(controller_->input_ref_.readFromNonRT()))->joint_names[0], joint_names_[0]);
+  EXPECT_FALSE(std::isnan((*(controller_->input_ref_.readFromNonRT()))->displacements[0]));
+  EXPECT_FALSE(std::isnan((*(controller_->input_ref_.readFromNonRT()))->velocities[0]));
+  EXPECT_FALSE(std::isnan((*(controller_->input_ref_.readFromNonRT()))->duration));
+  EXPECT_EQ((*(controller_->input_ref_.readFromNonRT()))->displacements[0], 0.45);
+  EXPECT_EQ((*(controller_->input_ref_.readFromNonRT()))->velocities[0], 0.0);
+  EXPECT_EQ((*(controller_->input_ref_.readFromNonRT()))->duration, 1.25);
```

Comment:
you should call this also a second time to be sure that this works only once - not you can not confirm that.

Reply:
extended the test here like [this](https://github.com/StoglRobotics/ros_team_workspace/commit/9692b171a48280a676e83c8cf021b6dfd315d703#diff-a73b5e0e5605b92a319fd503bf335487c5b3a833f0d4f39150e51a93bf566116R533-R539)   but have question,
it works when setting the ref_msg to nan like done [here](https://github.com/StoglRobotics/ros_team_workspace/commit/9692b171a48280a676e83c8cf021b6dfd315d703#diff-6e6a1012850b4b6040c5339b00ffad68f776d9edb72dbbe3e1d84482d52c4081R254-R256) 

but instead of individually setting them like that i wanted to just call 
reset_controller_reference_msg(*(input_ref_.readFromRT()), state_joints_, get_node());

but this is not working, also made sure that the fun is being called but the msg is not being reset. is the way i am inputing the ref_msg wrong? like this "*(input_ref_.readFromRT())".